### DISCUSSION TITLE
Using with NGINX

### DISCUSSION THREAD

**Discussion by jclusso** [basecamp/kamal#232] [General]
Was looking into MRSK and was wondering how using NGINX would work? Our application relies on NGINX in front since we have some Lua scripts. I was wondering if anyone has added NGINX yet? Seems like it would kinda be redundant to Traefik?

**Comment by jeremy** [comment]
We've been weaning ourselves off Nginx where we can, moving to app middleware, Traefik plugins, or load balancer and CDN features.

If you aren't using nginx as a reverse proxy to your app, you could explore running it as an accessory and routing specific paths to it with Traefik. If you are using it as a proxy, you'd need to extend MRSK to create and run images on a Docker network so your nginx container could proxy to the app using its container name.

Your best bet is to assess whether you can live without those Lua scripts or migrate them to another layer, like Traefik or a fronting CDN.

**Reply by jclusso** [reply]
Interesting, I’ll have to investigate if we could run them some other way. 

The purpose is really to offload a slow running request to poll Redis for job completion. We have a 30s timeout on specific API requests that run a worker that processes this and puts the result in Redis.

**Comment by MikeBinstock** [comment]
any update on how to use cloudflare to serve static parts? I don't really like to install nginx if I can serve all the assets directly from a CDN like cloudflare

**Reply by jclusso** [reply]
Not sure this is relevant. You can already use CloudFlare to serve static parts without NGINX.

**Reply by igor-alexandrov** [reply]
@PabloC we have 4 apps deployed on AWS with MRSK and our typical schema now to serve static parts is Rails -> AWS ALB -> AWS CloudFront. No need for Nginx at all. Seems like you can build the same solution with CloudFlare.

**Reply by MikeBinstock** [reply]
do you update your AWS CloudfFront assets at build time? is something you do in the DockerFile? Tks!

**Reply by igor-alexandrov** [reply]
No, CloudFront has ALB configured as an origin. In this case if any asset is missing in CloudFront (after new build has been deployed) it will be fetched through ALB from Rails.

**Comment by Splines** [comment]
In our case, we have video files that we want serve directly via nginx for better performance. To not loose the nice zero-downtime deploys, it'd be nice if Kamal could support multiple web containers: one that is running a web server like nginx, and one with the main app, e.g. a Rails server. Then one of them could be configured to proxy through all the requests, e.g. the nginx web server. This would allow for complex setups since we could customize the nginx server however we like.

Or let Kamal route all the traffic to an accessory. However, accessories are excluded from zero-downtime deploys, so maybe the first option would be better.

Here is the architecture I envision depicted for the state where Kamal performs the health-check for a new deploy. It seems that with Kamal `2.10.0` this is currently not possible.

![kamal setup](https://github.com/user-attachments/assets/7aa720ca-f82a-401f-8871-9ca96a539a20)

**Reply by Splines** [reply]
And here are some points why we'd love to have this feature in Kamal:
- Serve static video files that should be authorized by Rails (check for user role) via `X-Accel-Redirect` in nginx, see also: https://github.com/basecamp/kamal/discussions/820
- Serve static assets (JS/CSS) directly by nginx from disk without having to go through the Puma web server. This way, we can free up resources.

The assumption is that having to go through Rails is a lot slower. Maybe that has changed over the years, but don't web servers still excel at the task of serving static assets? Happy to learn about alternative, easy-to-configure setups (yes, nginx config files are a pain to manage).

Related:
- https://github.com/basecamp/kamal/discussions/1216
- https://github.com/basecamp/kamal/issues/1278

**Reply by Splines** [reply]
I've opened an issue for this here: https://github.com/basecamp/kamal/issues/1740

**Reply by tbuehlmann** [reply]
[Thruster](https://github.com/basecamp/thruster) deals with assets caching and supports X-Sendfile, maybe that's something?
